krun/util.py

```diff
@@ -192,16 +207,28 @@ def read_popen_output_carefully(process, platform, print_stderr=True):
                 if stderr_consumer is not None:
                     stderr_consumer.send(d)
 
+        if timeout_at is not None and datetime.now() >= timeout_at:
+            raise ExecutionTimeout()
+
     # We know stderr and stdout are closed.
     # Now we are just waiting for the process to exit, which may have
     # already happened of course.
+    while True:
+        process.poll()
+        if process.returncode is not None:
+            break
+
+        if timeout_at is not None:
+            debug("exit: %s" % datetime.now())
+        if timeout_at is not None and datetime.now() >= timeout_at:
+            raise ExecutionTimeout()
+        time.sleep(0.1)  # A small sleep to prevent us hogging too much CPU.
```

Comment:
Wouldn't it be better on the `poll`/`wait` to specify the remaining timeout and leave it to the OS to do this?

Reply:
That's what I proposed in the PR description I think.